base/applications/mspaint/mouse.cpp

```diff
@@ -58,216 +61,444 @@ BOOL nearlyEqualPoints(INT x0, INT y0, INT x1, INT y1)
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_BRUSH
+struct BrushTool : ToolBase
+{
+    BrushTool() : ToolBase(TOOL_BRUSH)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Brush(m_hdc, last.x, last.y, x, y, m_fg, toolsModel.GetBrushStyle());
+        else
+            Brush(m_hdc, last.x, last.y, x, y, m_bg, toolsModel.GetBrushStyle());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        draw(bLeftButton, x, y);
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_AIRBRUSH
+struct AirBrushTool : ToolBase
+{
+    AirBrushTool() : ToolBase(TOOL_AIRBRUSH)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Airbrush(m_hdc, x, y, m_fg, toolsModel.GetAirBrushWidth());
+        else
+            Airbrush(m_hdc, x, y, m_bg, toolsModel.GetAirBrushWidth());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        draw(bLeftButton, x, y);
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_TEXT
+struct TextTool : ToolBase
+{
+    TextTool() : ToolBase(TOOL_TEXT)
+    {
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        POINT temp;
+        if (bLeftButton)
+        {
+            imageModel.ResetToPrevious();
+            temp.x = max(0, min(x, imageModel.GetWidth()));
+            temp.y = max(0, min(y, imageModel.GetHeight()));
+            selectionModel.SetSrcAndDestRectFromPoints(start, temp);
+            RectSel(m_hdc, start.x, start.y, temp.x, temp.y);
+        }
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
```

Comment:
(Same comment regarding mouse-left or mouse-right for creating text zone)

Reply:
OK.